### PR TITLE
feat: poc for useId support

### PR DESCRIPTION
**Pull request by aniamusial**
# Related issue

https://vsf.atlassian.net/browse/SFUI2-563?atlOrigin=eyJpIjoiYmZhZDRhZGI5OTBkNDY1Y2IzMWQ5NDhkYWUzZThkYTAiLCJwIjoiaiJ9

# Reasons behind decision

Current implementation with generateId cause mismatch in react for a11y attributes of combobox and input.

Why useId? TLDR

The primary benefit of useId is that React ensures that **it works with server rendering.** 
IMPORTANT: It should not be used as key for arrays though! For such attributes we still can use generateId or index. 
useId is a recommended solution for **accessibility** fe. instead of:
`const listboxId = generateId('listbox');` 
we should use:
`const listboxId  = useId()`;

Unique ids are very difficult to guarantee with an incrementing counter because the order in which the client components are hydrated may not match the order in which the server HTML was emitted. By calling useId, you ensure that hydration will work, and the output will match between the server and the client.

Inside React, useId is generated from the “parent path” of the calling component. This is why, if the client and the server tree are the same, the “parent path” will match up regardless of rendering order.

# Other notes and references

For a generic solution for both frameworks there is also a possibility to use `uuid`  (https://github.com/uuidjs/uuid). 

The` generateId.ts` would like this:
`import { v4 as uuidv4 } from 'uuid';
let id = uuidv4();
export const generateId = (prefix: string) => 'vsf-${prefix}--${id}';
`
It provides unique id and is very unlikely to cause bugs. Still, it is not as bulletproof as `useId` with its conflict-handling. 

Therefore I would recommend to implement `uuid `as part of of generateId function that we will use for keys in arrays for both frameworks and use useId for accessibility attributes in react. 

**EDIT:** 
We can use Vue getCurrentInstance to create our own useId. 

`export const useId = () => getCurrentInstance().uid;`

# Discussion outcome

1. how about bundle size when using useId - how much it weights
Its a react hook. Size is negligible.

2. what about vue part, we need something similiar to mitigate issue of new generation of id there, also point 1 is applicable here as well
Discussed above - we will use getCurrentInstance from internal API.

3. maybe if we use something from outside we should wrap it up with out method, because (not sure tho if we need)
a. we have layer of abstraction when we will decide to change it somewhere in future?
b. unify naming?

Updated the PoC with following suggestions.

# Sources
https://www.webtips.dev/webtips/react-hooks/useid
https://beta.reactjs.org/reference/react/useId

# Checklist

- [ ] Self code-reviewed
- [ ] Changes documented
- [ ] Semantic HTML
- [ ] SSR-friendly
- [ ] Caching friendly
- [ ] a11y for WCAG 2.0 AA
- [ ] examples created
- [ ] created cypress tests

### CODE REVIEW

**Comment by dkacper** [comment]
I agree we should use React's `useId`. As mentioned, it solves SSR issues with generated IDs, and also it keeps id stable, while out generateId generates new ID per render.
Just to be noticed: useId docs https://beta.reactjs.org/reference/react/useId

**Comment by Szymon-dziewonski** [comment]
Questions to be answered:
1. how about bundle size when using `useId` - how much it weights
2. what about vue part, we need something similiar to mitigate issue of new generation of id there, also point 1 is applicable here as well
3. maybe if we use something from outside we should wrap it up with out method, because (not sure tho if we need)
   a. we have layer of abstraction when we will decide to change it somewhere in future? 
   b. unify naming?


Source code [useId](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useId.ts) kind seems reasonable. So I would say weight is neglectable.

However we need to find something similiar for vue, I cant find any specific alternative, so I would say we should extend/fix/modify what we already have, we can use https://stackoverflow.com/a/74915162

**Comment by FRSgit** [comment]
Just two cents about Vue API:
If I remember correctly for Vue this utils should work similarily to React's `useId`:
```ts
import { getCurrentInstance } from 'vue';
const uid = getCurrentInstance().uid;
```

So in the end - maybe we should just create an `useId()` method that will live in `vue/shared` directory? Like:
```ts
export const useId = () => getCurrentInstance().uid;
```
And in React - just use what their api gives to us

**Comment by Szymon-dziewonski** [comment]
@FRSgit exactly, this is what I add in my previous comment https://stackoverflow.com/questions/74914835/is-there-an-equivalent-to-reacts-useid-hook-in-vue-js/74915162#74915162

**Comment by aniamusial** [comment]
@FRSgit @Szymon-dziewonski I see that getCurrentInstance was removed in vue 3 and considered an anti-pattern (https://github.com/vuejs/docs/issues/1422#issuecomment-1032120675). I'll try to look for some similar solution

**Comment by Szymon-dziewonski** [comment]
@aniamusial nice research, I would say then we should not use it at all, if its not stable. I just checked and for vue its working correctly because `generateId` is called only once when component is run. 

So I would say we just should  create new `useId` util in `vue/shared` and use `generateId` inside of it, just in case someone in react would like to use `generateId` method, and also if we would change `generateId` into `useId` there would be 2x `useId` in react, wdyt?

File: packages/sfui/frameworks/vue/shared/useId.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import { getCurrentInstance } from 'vue';
+
+export const useId = () => getCurrentInstance()?.uid.toString();
```
Comment (dkacper) [inline packages/sfui/frameworks/vue/shared/useId.ts:3]: According to this [thread](https://stackoverflow.com/questions/72209080/vue-3-is-getcurrentinstance-deprecated) `getCurrentInstance` is kinda deprecated and not stable to use. 🤔 Can we safely use it?

**Review by dkacper — APPROVED** [review]
I'm not entirely convinced to use `getCurrentInstance()` in Vue, but I'm leaving the decision to Vue experts. 😄

**Comment by FRSgit** [comment]
Yup, going for separate implementation for Vue is the way to go also in my opinion. Just make sure that it works in Nuxt SSR and we're good to go 👍

File: packages/sfui/frameworks/vue/shared/useId.ts (hunk 1 above)
Comment (Szymon-dziewonski) [inline packages/sfui/frameworks/vue/shared/useId.ts:3]: @aniamusial we all literally said we dont want to go with `getCurrentInstance` @FRSgit @dkacper

Comment (aniamusial) [reply packages/sfui/frameworks/vue/shared/useId.ts:3]: I understood the @FRSgit comment as approval, aspecially that it was commented after getCurrentInstance was applied. I can remove it and change the generateId naming in vue to be useId - is it ok for now?

Comment (FRSgit) [reply packages/sfui/frameworks/vue/shared/useId.ts:3]: Yup, need to implement something else here 😄

**Pull request merged**